warp/Network/Wai/Handler/Warp/HTTP2.hs

```diff
@@ -50,8 +50,23 @@ http2 conn ii addr settings readN app =
                 | Just E.ThreadKilled  <- E.fromException e -> return ()
                 -- killed by the local timeout manager
                 | Just T.TimeoutThread <- E.fromException e -> return ()
-                | otherwise -> S.settingsOnException settings (Just req) e
+                | otherwise -> do
+                    S.settingsOnException settings (Just req) e
             return ResponseReceived
+        case eResponseReceived of
```

Comment:
This essentially duplicates the `settingsOnException` logic: once for exceptions from the response, and once from the application itself. However, I'm not following the motivation for this: why do we need to `try` twice and handle these cases differently? At the very least, some comments explaining would be great.

Reply:
> However, I'm not following the motivation for this: why do we need to try twice and handle these cases differently? At the very least, some comments explaining would be great.

@snoyberg the motivation comes not from analyzing the code but from a real-world servant app that doesn't trigger `settingsOnException` when actions are throwing an exception. I've made a minimal reproduction repository with full instructions present in this PR.

Now, to your question as to why do we need to catch anything inside the action itself – I don't know, but since I didn't have time to be able to write a piece of code that would be able to go through that branch of code, I decided to keep the behavior as it is currently. Definitely feels like it could be eliminated, but I am not 100% confident on that. Any comments from @kazu-yamamoto would be very much appreciated!

Reply:
Separate from issues within `warp` or `warp-tls`, this sounds very reminiscent of https://github.com/haskell-servant/servant/issues/309.

Reply:
Yes, this was exactly the problem I started seeing when I've switched from default servant to a custom `AppM = ReaderT Env IO` thing to use it together with `UnliftIO` and rio stuff

> setOnException doesn't work for throwError in a Handler. It seems that if you throw an exception in IO (ie. liftIO $ throw someException), that gets handled by the default exception handler in warp. However, throwing a ServantErr doesn't get caught by warp. Servant responds with a 500 error, but nothing gets logged.

And this PR should fix it.